### PR TITLE
Add Validator Mock

### LINKED ISSUE #369: Create Mocked Validator
## 📚 Context
Now that we have a general design for the database schema and how the Validator will interact with it, it is time to exercise the component's main logic.

## ✔️ Solution
To create a prototype to validate the Validator's main logic and APIs, mocking the database and the proof-generating code.

## 📈 Subtasks
- [x] Create the necessary types
- [x] Implement the main logic
- [x] Test main logic with mocks

### PR DESCRIPTION
**Pull request by torives**
Closes #369

### CODE REVIEW

File: internal/node/model/models.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	// Output data as a blob, starting with '0x'
+	Blob hexutil.Bytes `json:"blob"`
+}
+
+type Epoch struct {
+	StartBlock uint64
+	EndBlock   uint64
+}
+
+type InputRange struct {
```
Comment (GMKrieger) [inline internal/node/model/models.go:33]: I have some doubts about the need to have an InputRange struct. It isn't a "thing" by itself in the architecture, it's a value related to the claim. In your mock, the use of the struct aids in appending the values, but also adds a lot of lines to access these values.

I don't really have an alternative, just food for thought.

File: internal/validator/validator.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+// generateProofs will create the proofs for all Outputs within an Epoch.
+// It returns the proofs and the root hash of the Merkle tree used to generate them
+func (v Validator) generateProofs(
+	ctx context.Context,
+	inputRange InputRange,
+	machineStateHash hexutil.Bytes,
+	outputs []Output,
+) ([]Proof, hexutil.Bytes, error) {
```
Comment (GMKrieger) [inline internal/validator/validator.go:196]: Do we need to return the OutputsEpochRootHash as a different value? Can't we get it from any of the proofs?

File: internal/validator/validator_test.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+		errChannel <- validator.Start(ctx, ready)
+	}()
+	cancel()
+
+	err := <-errChannel
+	s.Require().Equal(context.Canceled, err, "stopped for the wrong reason")
+}
+
+func (s *ValidatorSuite) TestItEventuallyBecomesReady() {
+	validator := NewValidator(s.repo, uint64(10), uint64(0))
```
Comment (GMKrieger) [inline internal/validator/validator_test.go:66]: To be consistent with the other tests

```
	epochDuration := uint64(10)
	inputBoxDeploymentBlock := uint64(0)
	validator := NewValidator(s.repo, epochDuration, inputBoxDeploymentBlock)
```

**Review by GMKrieger — COMMENTED** [review]
Just some minor comments

File: internal/validator/validator.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	repo ValidatorRepository,
+	epochDuration uint64,
+	inputBoxDeploymentBlock uint64,
+) Validator {
+	return Validator{repo, epochDuration, inputBoxDeploymentBlock}
+}
+
+// generateProofs will create the proofs for all Outputs within an Epoch.
+// It returns the proofs and the root hash of the Merkle tree used to generate them
+func (v Validator) generateProofs(
```
Comment (gligneul) [inline internal/validator/validator.go:191]: I would make this a stand-alone function and put it in another file.

File: internal/node/model/models.go (hunk 1 above)
Comment (torives) [reply internal/node/model/models.go:33]: These are valid points and I agree with you: it isn't needed in any way. 

The `InputRange` struct was a late addition born from my intention to create familiarity between the node's code and the on-chain contracts (at least regarding the main entities) and thus ease understanding. IMO the tradeoff is worth it.

File: internal/validator/validator.go (hunk 2 above)
Comment (torives) [reply internal/validator/validator.go:196]: We can. I'll remove it from the returned values.

**Review by gligneul — APPROVED** [review]

**Pull request merged**